File: CSharp/Configuration/ItemSetting.cs
using System;
using System.Collections;

namespace SecureSwitch.Configuration {

	/// <summary>
	/// Indicates the type of security for a file or directory.
	/// </summary>
	public enum SecurityType {
		/// <summary>
		/// The item should be secure.
		/// </summary>
		Secure,

		/// <summary>
		/// The item should be insecure.
		/// </summary>
		Insecure,

		/// <summary>
		/// The item should be ignored.
		/// </summary>
		Ignore
	}


	/// <summary>
	/// The ItemSettingComparer class implements the IComparer interface to compare.
	/// </summary>
	public class ItemSettingComparer : IComparer {
		/// <summary>
		/// Initialize an instance of this class.
		/// </summary>
		public ItemSettingComparer() {
		}

		/// <summary>
		/// Compares the two objects as string and ItemSetting or both ItemSetting 
		/// by the Path property.
		/// </summary>
		/// <param name="x">First object to compare.</param>
		/// <param name="y">Second object to compare.</param>
		/// <returns></returns>
		public int Compare(object x, object y) {
			// Check the type of the parameters
			if (!(x is ItemSetting) && !(x is string))
				// Throw an exception for the first argument
				throw new ArgumentException("Parameter must be a ItemSetting or a String.", "x");
			else if (!(y is ItemSetting) && !(y is string))
				// Throw an exception for the second argument
				throw new ArgumentException("Parameter must be a ItemSetting or a String.", "y");

			// Initialize the path variables
			string xPath = string.Empty;
			string yPath = string.Empty;

			// Get the path for x
			if (x is ItemSetting)
				xPath = ((ItemSetting)x).Path;
			else
				xPath = (string)x;

			// Get the path for y
			if (y is ItemSetting)
				yPath = ((ItemSetting)y).Path;
			else
				yPath = (string)y;

			// Compare the paths, ignoring case
			return string.Compare(xPath, yPath, true);
		}
	}


	/// <summary>
	/// The ItemSetting class is the base class that represents entries in the &lt;secureSwitch&gt;
	/// configuration section.
	/// </summary>
	public class ItemSetting {
		// Fields
		private SecurityType secure = SecurityType.Secure;
		private string path = string.Empty;

		/// <summary>
		/// Gets or sets the type of security for this directory or file.
		/// </summary>
		public SecurityType Secure {
			get { return secure; }
			set { secure = value; }
		}

		/// <summary>
		/// Gets or sets the path of this directory or file.
		/// </summary>
		public string Path {
			get { return path; }
			set { path = value; }
		}

		/// <summary>
		/// Creates an instance of this class.
		/// </summary>
		public ItemSetting() {
		}

		/// <summary>
		/// Creates an instance with initial values.
		/// </summary>
		/// <param name="path">The relative path to the directory or file.</param>
		/// <param name="ignore">A flag to ignore security for the directory or file.</param>
		public ItemSetting(string path, SecurityType secure) {
			// Initialize the path and secure properties
			this.path = Path;
			this.secure = secure;
		}

		/// <summary>
		/// Creates an instance with an initial path value.
		/// </summary>
		/// <param name="path">The relative path to the directory or file.</param>
		public ItemSetting(string path)
			: this(path, SecurityType.Secure) {
		}

	}

	/// <summary>
	/// The ItemSettingCollection class houses a collection of ItemSetting instances.
	/// </summary>
	public class ItemSettingCollection : CollectionBase {

		/// <summary>
		/// Initialize an instance of this collection.
		/// </summary>
		public ItemSettingCollection() {
		}

		/// <summary>
		/// Returns the index of a specified item in the collection.
		/// </summary>
		/// <param name="item">The item to find.</param>
		/// <returns>Returns the index of the item.</returns>
		public int IndexOf(ItemSetting item) {
			return List.IndexOf(item);
		}

		/// <summary>
		/// Returns the index of an item with the specified path in the collection.
		/// </summary>
		/// <param name="Path">The path of the item to find.</param>
		/// <returns>Returns the index of the item with the path.</returns>
		public int IndexOf(string path) {
			// Create a comparer for sorting and searching
			ItemSettingComparer Comparer = new ItemSettingComparer();
			InnerList.Sort(Comparer);
			return InnerList.BinarySearch(path, Comparer);
		}

	}

}

File: VB/Configuration/ItemSetting.vb
Imports System
Imports System.Configuration
Imports System.Globalization
Imports System.Text.RegularExpressions

Namespace SecureSwitch.Configuration

	''' <summary>
	''' Indicates the type of security for a file or directory.
	''' </summary>
	Public Enum SecurityType

		''' <summary>
		''' The item should be secure.
		''' </summary>
		Secure

		''' <summary>
		''' The item should be insecure.
		''' </summary>
		Insecure

		''' <summary>
		''' The item should be ignored.
		''' </summary>
		Ignore

	End Enum

	''' <summary>
	''' Represents an file or directory entry in the &lt;secureSwitch&gt;
	''' configuration section.
	''' </summary>
	Public MustInherit Class ItemSetting
		Inherits ConfigurationElement

#Region "Constructors"

		''' <summary>
		''' Creates an instance of ItemSetting.
		''' </summary>
		Public Sub New()
			MyBase.New()
		End Sub

		''' <summary>
		''' Creates an instance with an initial path value.
		''' </summary>
		''' <param name="path">The relative path to the file.</param>
		Public Sub New(ByVal path As String)
			Me.New()
			Me.Path = path
		End Sub

		''' <summary>
		''' Creates an instance with initial values.
		''' </summary>
		''' <param name="path">The relative path to the file.</param>
		''' <param name="secure">The type of security for the file.</param>
		Public Sub New(ByVal path As String, ByVal secure As SecurityType)
			Me.New(path)
			Me.Secure = secure
		End Sub

#End Region

#Region "Properties"

		''' <summary>
		''' Gets or sets the path of this item.
		''' </summary>
		<ConfigurationProperty("path", IsRequired:=True, IsKey:=True)> _
		Public Overridable Property Path() As String
			Get
				Return CStr(Me("path"))
			End Get
			Set(ByVal value As String)
				Me("path") = value
			End Set
		End Property

		''' <summary>
		''' Gets or sets the type of security for this item.
		''' </summary>
		<ConfigurationProperty("secure", DefaultValue:=SecurityType.Secure)> _
		Public Property Secure() As SecurityType
			Get
				Return CType(Me("secure"), SecurityType)
			End Get
			Set(ByVal value As SecurityType)
				Me("secure") = value
			End Set
		End Property

#End Region

	End Class

	''' <summary>
	''' Represents a collection of ItemSetting objects.
	''' </summary>
	Public MustInherit Class ItemSettingCollection
		Inherits ConfigurationElementCollection

		''' <summary>
		''' Returns the index of a specified item in the collection.
		''' </summary>
		''' <param name="item">The item to find.</param>
		''' <returns>Returns the index of the item.</returns>
		Public Function IndexOf(ByVal item As ItemSetting) As Integer
			If Not item Is Nothing Then
				Return BaseIndexOf(item)
			Else
				Return -1
			End If
		End Function

		''' <summary>
		''' Returns the index of an item with the specified path in the collection.
		''' </summary>
		''' <param name="Path">The path of the item to find.</param>
		''' <returns>Returns the index of the item with the path.</returns>
		Public Function IndexOf(ByVal path As String) As Integer
			If path Is Nothing Then
				Throw New ArgumentNullException("path")
			Else
				Return Me.IndexOf(CType(BaseGet(path.ToLower(CultureInfo.InvariantCulture)), ItemSetting))
			End If
		End Function

	End Class

End Namespace